flocker/control/_protocol.py

```diff
@@ -856,11 +1003,156 @@ def logger(self):
+            of the new configuration.
+        :param state: The new state.
+        :param state_generation: The expected resulting generation hash of the
+            new state.
+        """
+        self._set_configuration(configuration, configuration_generation)
+        self._set_state(state, state_generation)
+        self._update_agent()
+
     @ClusterStatusCommand.responder
-    def cluster_updated(self, eliot_context, configuration, state):
+    def cluster_updated(
+            self, eliot_context, configuration, configuration_generation,
+            state, state_generation
+    ):
+        """
+        Responder to ``ClusterStatusCommand``. Updates the configuration and
+        state to the passed in values, and validates the hashes.
+
+        :param eliot_context: The eliot context this is called under.
+        :param configuration: The new configuration.
+        :param configuration_generation: The expected generation hash of the
+            new configuration.
+        :param state: The new state.
+        :param state_generation: The expected generation hash of the new state.
+        """
         with eliot_context:
-            self.agent.cluster_updated(configuration, state)
-            return {}
+            self._update_cluster(configuration, configuration_generation,
+                                 state, state_generation)
+            return self._current_generations_response()
+
+    @ClusterStatusDiffCommand.responder
+    def cluster_updated_diff(
+            self, eliot_context,
+            configuration_diff,
+            start_configuration_generation,
+            end_configuration_generation,
+            state_diff,
+            start_state_generation,
+            end_state_generation,
+    ):
+        """
+        Responder to ``ClusterStatusDiffCommand``. Updates the configuration
+        and state by applying diffs to the current values, and verifies the
+        hash of the resulting objects.
+
+        :param eliot_context: The eliot context this is called under.
+        :param configuration_diff: The diff from the current configuration to
+            the next one.
+        :param start_configuration_generation: The expected generation hash of
+            former configuration known by this node.
+        :param end_configuration_generation: The expected generation hash of
+            the new configuration after the diff is applied.
+        :param state: The new state.
+        :param start_state_generation: The expected generation hash of former
+            state known by this node.
+        :param end_state_generation: The expected generation hash of the new
+            state after the diff is applied.
+        """
+        with eliot_context:
+            if (start_configuration_generation !=
+                    self._current_configuration_generation):
+                return self._current_generations_response()
```

Comment:
:question: So if the agent configuration generation has changed since it last reported its hashes to the control service  (somehow??) we return the latest hashes.
The control service should then send us an updated diff next time round.